service/dap/server.go

```diff
@@ -1011,40 +1011,228 @@ func (s *Server) onSetBreakpointsRequest(request *dap.SetBreakpointsRequest) {
 
 	// According to the spec we should "set multiple breakpoints for a single source
 	// and clear all previous breakpoints in that source." The simplest way is
-	// to clear all and then set all.
+	// to clear all and then set all. To maintain state (for hit count conditions)
+	// we want to amend existing breakpoints.
 	//
-	// TODO(polina): should we optimize this as follows?
 	// See https://github.com/golang/vscode-go/issues/163 for details.
 	// If a breakpoint:
 	// -- exists and not in request => ClearBreakpoint
 	// -- exists and in request => AmendBreakpoint
 	// -- doesn't exist and in request => SetBreakpoint
 
-	// Clear all existing breakpoints in the file that are not function breakpoints.
-	if err := s.clearMatchingBreakpoints(func(bp *api.Breakpoint) bool { return bp.File == serverPath && bp.Name == "" }); err != nil {
+	// Get all existing breakpoints that match for this source.
+	sourceRequestPrefix := fmt.Sprintf("sourceBp Path=%s", request.Arguments.Source.Path)
+	existingBps := s.getMatchingBreakpoints(sourceRequestPrefix)
```

Comment:
I liked the old way of finding breakpoints better. This doesn't even work if a file's path happens to be the prefix of another file's path.

Reply:
That makes sense. I can change the finding breakpoint logic back (or can use %q and add a " Line=" after Path which I think should resolve the Prefix issue

One thing I was thinking though for a method like this, is that we are guaranteed to only get breakpoints set by the DAP server, so if we eventually support parallel clients the DAP client won't clear all breakpoints set by an RPC user.

Reply:
@aarzilli I assume by your LGTM this has been resolved?

Reply:
Yes.